Fully evented request handler implementation

weierophinney opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

Building on #24, this patch provides a fully-evented `SwooleRequestHandlerRunner` implementation.

It extracts the body of each of the `on*()` methods of that class into dedicated listeners.
The original methods now marshal a typed event class (see the `Mezzio\Swoole\Event` namespace for event types), using the arguments provided by the Swoole HTTP server when dispatching them; they then dispatch the event class via a [PSR-14 event dispatcher](https://www.php-fig.org/psr/psr-14/).
As such, the runner now only composes the HTTP server instance and the PSR-14 event dispatcher in order to do its work.

In some cases, multiple listeners were extracted:

- For the workerstart event, we have the general workerstart logic in one listener, as well as another that performs hot code reloading.
- For the request event, we have a listener for static resource handling, as well as one for dispatching a PSR-15 application.
- For the task event, we have a TaskInvokerListener and a TaskEventDispatcherListener (copious documentation and examples provided for these, as the workflow is somewhat complex)

A listener provider for composing listeners is part of the patch, as well as a factory that uses configuration in order to seed it.
This approach allows users to append new listeners, or, if they are adventurous, use the laminas-stdlib `MergeReplaceKey` class to reset the listeners to a specified workflow.

Fixes #1

basz commented:
Whilst testing this it seems the onTask handler is required... (Needed to upgrade to swoole 4.4 -> 4.5 also, so might be due to that)

```
PHP Fatal error:  Swoole\Server::start(): failed to start server. Error: start_check: onTask event callback must be set in /app/vendor/mezzio/mezzio-swoole/src/SwooleRequestHandlerRunner.php on line 65
```

```
$this->httpServer->on('task', [$this, 'onTask']);
```

```
    /**
     * Handle tasks
     */
    public function onTask(SwooleHttpServer $server, int $taskId, int $fromWorkerId, mixed $data): void
    {
        // tbw
    }
```

I'm guessing a finish handler might also be required then

weierophinney commented:
@basz I've pushed my latest changes, and they include the task and finish handlers.

Interestingly, I discovered the finish handler is not required, nor are you required to call `$server->finish()` when done processing a task, at least as of Swoole 4.5. I'm not sure why, exactly.

Tests are failing, but that's because I haven't updated them; I've been busy trying to get the task features to work correctly, primarily. I'll have more to push likely early next week.

basz commented:
The finish thing is a little awkward... it currently *is* required to start the server, but each `server->task()` has an finish handler as third parameter... But perhaps it may be useful if the Worker responsible for dispatching tasks can keep track of running tasks, do it can queue them when not free TasWorkers re available... dunno.

Would you consider adding a 'RecieveEvent'? It would enable TaskWorkers (which my not call `$server->task(...)` themselves) to initiate tasks by sending them back to a worker via a client...

Not sure if this is even possible with the HTTP server process. Just tested that the client can connect (to port 80) and do a send but the servers onRecieve call doesn't seem to be called. hmmm...

https://developpaper.com/example-of-implementing-real-time-asynchronous-task-queue-with-php-extended-swoole/

weierophinney commented:
@basz I've just pushed documentation and a number of code changes, all related to task handling; please check when you can.

Last item on the TODO is tests for each of these new classes, and then this should be ready.

froschdesign commented on mkdocs.yml:
```diff
@@ -2,12 +2,26 @@ docs_dir: docs/book
+        - "Command Line Tooling": v3/command-line.md
+        - "How it works": v3/how-it-works.md
+        - "Considerations when using Swoole": v3/considerations.md
+        - Migration: v3/migration.md
```
```suggestion
        - Migration:
            - "Migration from Version 2 to 3": migration.md
```

froschdesign commented on mkdocs.yml:
```diff
@@ -2,12 +2,26 @@ docs_dir: docs/book
     - Home: index.md
+    - v3:
+      - Introduction: v3/intro.md
+      - Reference:
+        - "Swoole HTTP Server Events": v3/events.md
+        - "Static Resources": v3/static-resources.md
+        - Logging: v3/logging.md
+        - "Async Tasks": v3/async-tasks.md
+        - "Using Swoole Tables": v3/table.md
+        - "Hot Code Reloading": v3/hot-code-reload.md
+        - "Command Line Tooling": v3/command-line.md
+        - "How it works": v3/how-it-works.md
+        - "Considerations when using Swoole": v3/considerations.md
```
```suggestion
      - "Swoole HTTP Server Events": v3/events.md
      - "Static Resources": v3/static-resources.md
      - Logging: v3/logging.md
      - "Async Tasks": v3/async-tasks.md
      - "Using Swoole Tables": v3/table.md
      - "Hot Code Reloading": v3/hot-code-reload.md
      - "Command Line Tooling": v3/command-line.md
      - "How it works": v3/how-it-works.md
      - "Considerations when using Swoole": v3/considerations.md
```

froschdesign commented on docs/book/v3/events.md:
````diff
@@ -0,0 +1,251 @@
+    ],
+    // ...
+];
+```
````
I think this part can be extracted as cookbook recipe and then linked from this page.

froschdesign commented on docs/book/v3/events.md:
```diff
@@ -0,0 +1,251 @@
+To make the system more flexible, version 3 of this library now has the `Mezzio\Swoole\SwooleRequestHandlerRunner` class compose a [PSR-14 EventDispatcherInterface](https://www.php-fig.org/psr/psr-14/) instance that is then triggered for each Swoole HTTP Server event.
+Additionally, the arguments passed to the Swoole listener are now aggregated into typed event classes, allowing access to the arguments by your PHP listener classes.
+
+## Event classes
```
```suggestion
## Event Classes
```

froschdesign commented on docs/book/v3/events.md:
```diff
@@ -0,0 +1,251 @@
+> As such, if you want other listeners on that event, they **must** resolve before that listener.
+> An example of such a listener is the `Mezzio\Swoole\Event\StaticResourceRequestListener`, which only stops propagation if it was able to resolve the request to a static resource.
+
+## Providing a dispatcher and listeners
```
```suggestion
## Providing a Dispatcher and Listeners
```

froschdesign commented on docs/book/v3/events.md:
````diff
@@ -0,0 +1,251 @@
+    /**
+     * "Return value" of the task; what to pass to finish()
+     *
+     * @param mixed $returnValue
+     */
+    public function setReturnValue($returnValue): void;
+
+    /** @return mixed */
+    public function getReturnValue();
+
+    /**
+     * Call this to indicate processing is complete, and to stop processing
+     * further listeners.
+     */
+    public function taskProcessingComplete(): void;
+}
+
+/**
+ * Describes "finish" event, called when processing a task has been completed
+ */
+class TaskFinishEvent
+{
+    public function getServer(): SwooleHttpServer;
+    public function getTaskId(): int;
+
+    /**
+     * Data passed to the finish() method
+     */
+    public function getData(): mixed;
+}
+
+
+/**
+ * Describes "shutdown" event
+ */
+class ServerShutdownEvent
+{
+    public function getServer(): SwooleHttpServer;
+}
+```
````
I would suggest splitting this code block. Each event is listed with a headline and a code block. For example:

``````
### Shutdown Event

```php
namespace Mezzio\Swoole\Event;

use Swoole\Http\Server;

class ServerShutdownEvent
{
    public function getServer(): Server;
}
```

froschdesign reviewed (changes requested):
I added some comments for the new events section in the documentation. the navigation can also be simplified.

weierophinney commented:
@froschdesign I've committed your suggestions, and also:

- Created a "cookbook" section, with the FAQ item from the events chapter, as suggested. (We've used the term "Cookbook" elsewhere, including the main Mezzio manual.)
- Split the event interfaces into individual h3-headed sections, as suggested.

I also moved the "Migration" heading out from under reference, to make it a peer with "Reference" and "Cookbook"

basz commented:
When couroutine support for tasks is enabled the signature for the onTask method changes as noted https://docs.mezzio.dev/mezzio-swoole/v2/async-tasks/#registering-the-handlers.

This PR currently fails to address this.

when coroutine enabled

```
$this->httpServer->on('task', [$this, 'onTaskWithCoroutineSupport']);
```

else

```
$this->httpServer->on('task', [$this, 'onTask']);
```


```
public function onTaskWithCoroutineSupport(SwooleHttpServer $server, Task $task)
{
    return $this->onTask($server, $task->id, $task->worker_id, $task->data);
}
```

weierophinney commented:
> When couroutine support for tasks is enabled the signature for the onTask method changes as noted https://docs.mezzio.dev/mezzio-swoole/v2/async-tasks/#registering-the-handlers.
> 
> This PR currently fails to address this.

Thanks for reminding me about this - I'll get a change pushed shortly!

weierophinney commented:
@basz Fixed now!

basz commented on src/SwooleRequestHandlerRunner.php:
```diff
@@ -185,4 +227,27 @@ public function onShutdown(SwooleHttpServer $server): void
+        return new TaskEvent($server, $taskId, $workerId, $data);
+    }
+
+    private function createTaskEventFromTaskObject(SwooleHttpServer $server, object $task): TaskEvent
```
Isn't that simply a \Swoole\Server\Task instance?

weierophinney replied:
Had to hunt quite a bit to find this class defined - it's not documented!

There's one issue I have with using it as a typehint: the class is marked final, so there's no way for me to mock it. And there's no way to instantiate it with the data necessary. As such, I'm going to leave this as-is, as it allows me to assert the structure and behavior required.

basz commented on docs/book/v3/async-tasks.md:
```diff
@@ -0,0 +1,888 @@
+// implementation, and defines a `listen()` method, and $container is a PSR-11
+// container implementation:
+$listenerProvider = $factory();
+$listener->listen(UserCreationEvent::class, $container->get(UserCreationListener::class));
```
replace `$listener->listen(` with `$listenerProvider->listen(`?

basz commented on docs/book/v3/async-tasks.md:
````diff
@@ -0,0 +1,888 @@
+```php
+namespace Example;
+
+class DeferrableTaskListener
````
class SomeDeferrableTaskListener

basz commented on docs/book/v3/async-tasks.md:
```diff
@@ -0,0 +1,888 @@
+
+    public function __invoke(SomeDeferrableEvent $event): void
+    {
+        $this->log->info(sprintf('Message: %s', (string) $event));
```
$this->logger

basz commented on docs/book/v3/async-tasks.md:
````diff
@@ -0,0 +1,888 @@
+    new SomeDeferrableEvent('hello world')
+);
+```
+
````
Confused as to why a SomeDeferrableEvent with a *listener* is created, and not a concrete SomeDeferrableTask invoked by the TaskInvokerListener?

weierophinney replied:
This is a lead up to the next section, where the code that creates it will be _dispatching an event_, not creating a task.

Basically, this whole write-up is leading up to the idea that instead of manually and explicitly queuing tasks with the server, you will dispatch events, and have listeners opt-in to whether they execute asynchronously via task workers by adding the `DeferredListenerDelegator` or `DeferredServiceListenerDelegator`.

basz commented on docs/book/v3/async-tasks.md:
```diff
@@ -0,0 +1,888 @@
+
+## Shipped task listeners
+
+This package ships two listeners that you can use to process tasks: `Mezzio\Swoole\Task\TaskEventDispatchListener` and `Mezzio\Task\TaskInvokerListener`.
```
Please explain a bit about the purpose and differences between TaskEventDispatchListener and TaskInvokerListener. Can or should I bring my own Listener to execute Tasks? Can I use them both or do I have to choose?

I am wondering why a simple `server->task(MyObjectRepresentingATask::createWith($oayload))` and a simple worker listener capable of pulling the task service (MyObjectRepresentingATask::class) from the service manager and invoking it with the given payload isn't provided. You must think that is a bad idea?

weierophinney replied:
There's actually three ways to handle tasks:

- Register a listener on the `TaskEvent`, and handle the `TaskEvent` directly. This requires introspecting the result of `$event->getData()` to determine if you can handle it in the first place. As such, it tends to be less type-safe, and thus more brittle.

- Use the `TaskEventDispatchListener`. In this case, the `SwooleRequestHandlerRunner::onTask()` method will invoke that listener, which pulls the data from the payload and treats it as an event to pass to the PSR-14 event dispatcher. You then register listeners agaisnt the various events that will be used as tasks. This is more type safe.

The above two **require** that you queue tasks using `$server->task()`, which means composing the Swoole HTTP Server instance in every class that needs to queue a task. I personally think that's a bad idea.

Why? Because I don't feel your handlers or middleware or services should be aware of the SAPI being used.

So, that leads us to the last approach, the `TaskInvokerListener`. With this one, when you queue a task, you create a `Task` instance that composes both the payload, and the code that will handle it. The `TaskInvokerListener` then checks if `$event->getData()` is a `TaskInterface` instance, and, if so, invokes it.

You _can_ create the task instance manually and pass it to `$server->task()`. But this patch provides an alternate solution: decorating your listener in either `DeferredListener` or `DeferredServiceListener` instances, via the corresponding delegator factories. What does this achieve?

It means you can compose a PSR-14 event dispatcher in your classes instead, and just dispatch events. If a listener can be executed asynchronously, then, in your configuration, you add one of the above delegators to it, and the delegator takes care of creating the appropriate task instance and passing it to the server.

A `DeferredListener` works fine for listeners that have no constructor dependencies or use no objects or resources as dependencies. The `DeferredServiceListener` is good for all the others. Tasks receive the container as an argument when invoked, and the `ServiceBasedTask` will then use that to pull the actual listener from the container and use it.

Now, as for your question of whether you are limited to one listener for tasks or can use multiple: you can use as many as you want. The `TaskEventDispatchListener` assumes that if the payload is an object, it can handle it, and will stop propagation. Similarly, the `TaskInvokerListener` assumes that if it receives a `TaskInterface` in the payload, it can handle it, and will stop propagation.

Personally, though, I'd stick to the `TaskInvokerListener`, and even recommend that later in this particular document.

basz reviewed (commented):
I see an architecture where events are centric. This is probably good, but might be difficult to understand as it significantly adds complexity. This PR does not simply translates swoole events into first-class citizens but seems to aim at an event-driven architecture. One where dispatching events will trigger async tasks.

I like it cause learning, but pwew... :-) thanks for the work done at yet another level...

basz commented:
Could it be that the HotCodeReloaderWorkerStartListenerFactory is missing a `$fileWatcher->addFilePath(\getcwd());` and therefore not working?

as https://github.com/mezzio/mezzio-swoole/blob/aa6fa2fbf28a820acee1d15b587fc67a8606d42f/src/HotCodeReload/Reloader.php#L90 has been removed

weierophinney commented:
> Could it be that the HotCodeReloaderWorkerStartListenerFactory is missing a `$fileWatcher->addFilePath(\getcwd());` and therefore not working?

If you want the working dir to be scanned, then you'd need to specify that as a path to the file watcher. You can do that currently via a delegator factory; I'll add configuration for the InotifyFileWatcher and/or HotCodeReloaderWorkerStartListenerFactory consume.

(BTW, tested this, and it hot code reloading definitely works now!)

weierophinney commented:
> > Could it be that the HotCodeReloaderWorkerStartListenerFactory is missing a `$fileWatcher->addFilePath(\getcwd());` and therefore not working?
> 
> If you want the working dir to be scanned, then you'd need to specify that as a path to the file watcher. You can do that currently via a delegator factory; I'll add configuration for the InotifyFileWatcher and/or HotCodeReloaderWorkerStartListenerFactory consume.
> 
> (BTW, tested this, and it hot code reloading definitely works now!)

@basz — Updated with ~~2a77773~~ de298dd!

weierophinney commented:
@basz —

> I see an architecture where events are centric. This is probably good, but might be difficult to understand as it significantly adds complexity. This PR does not simply translates swoole events into first-class citizens but seems to aim at an event-driven architecture. One where dispatching events will trigger async tasks.

The patch is trying to accommodate a few things:

- Make it possible for third-party packages and the application to tie into Swoole events without clobbering each other. Before, depending on the event, multiple listeners could attempt to register with the Swoole HTTP Server instance, which could either lead to silently doing something you don't expect, or raising an error during bootstrap. Neither was optimal.
- Simplify the logic of the HTTP handler runner implementation. Previously, it has a lot of branching logic, and we had to add new methods each time we wanted to handle new events. We now handle all related server events, and the class just registers event handlers and dispatches events.
- Allow triggering async tasks transparently, which builds off the previous two.

When I've presented on Swoole, I've noted that the two driving factors for using an async runtime are _parallelization_ ) and _deferment of execution_. The former happens naturally when you have an event loop and multiple workers. The latter is a result of using async features in the engine, whether those are coroutines, promises, async/await, or message queues. Swoole gives us one implicit way to benefit from async (coroutine support) and two ways to explicitly make functionality asynchronous: `defer()` (which has a ton of problems I've found) and `task()` (which acts as a message queue). Since tasks are part of the server event system, and they provide execution deferment, it made sense to accommodate them in this patch as well.

The main issue I have with any async runtime, though, is when I **explicitly** use those async features, such as `$server->task()`. Why?

- They typically make me hard-code to the specific async implementation. If I want to switch implementations... I have to rewrite/refactor.
- If I find I need to debug, or need to potentially move to a non-async architecture... I'm hard-coded to use async, and have to rewrite/refactor.

To my mind, if we are doing something async where it can be "fire and forget", being able to de-couple that from the async engine makes sense, and that's why I introduced the `Deferred*Listener` stuff + the `TaskInvokerListener`. These allow dispatching standard PSR-14 events in our code, and we can, in configuration, decorate our listeners to opt-in to the async runtime, _without needing to change how the code works_.

(Interestingly, because of the nature of the task support in Swoole 4.5, we can call tasks _synchronously_ as well — and because listeners can _update_ the event, I made the task listener _return the event_... which means you benefit from them in this way as well!)

In practice, this means writing standard listeners, and then just adding a delegator factory to decorate listeners we want to run via tasks. So, while the _internals_ are somewhat complex, the _interface exposed to the consumer_ is simpler. That was my explicit goal.
